packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -1242,6 +1242,12 @@ class TransformLayer extends OffsetLayer {
 /// When debugging, setting [debugDisableOpacityLayers] to true will cause this
 /// layer to be skipped (directly replaced by its children). This can be helpful
 /// to track down the cause of performance problems.
+///
+// Try to avoid an [OpacityLayer] with no children. Painting an [OpacityLayer]
+// in the Flutter engine is very costly due to the creation of an offscreen
+// buffer and render target switches. If there's no child, having the
+// OpacityLayer or not has the same effect. If [OpacityLayer] has no child, its
```

Comment:
This seems off to me.  Why can't we make it cheap in the engine if there's no child?  E.g. if the engine gets an OpacityLayer with no layers, it's a no-op layer.

My understanding is that it's still beneficial to omit such layers from the engine side because it would (slightly) shrink the size of the layer tree and thus the real time it takes to traverse it.

Reply:
Do you mean "Try to avoid an [OpacityLayer] with no children. Remove that layer if possible to save some tree walks." ?

I guess that we can make an empty OpacityLayer as a no-op in the engine. We're just not currently doing it, and instead doing FML_DCHECK the nonemptiness.

Reply:
I think it's fine for the contract to be `Don't send me an empty opacity layer`, but the explanation here makes me wonder why that is (e.g. why is it so easy to short-circuit it here and not in the engine?).  

Should we be pushing the short-circuiting even further up the call stack to the `pushLayer` call with opacity?  Should we add an assert to `OpacityLayer` that it has at least one child when it's pushed?

Reply:
Talked offline - let's make this a doc comment (`///`) and your new text in there should be good.

Reply:
Doc updated.